Multi-node support for cuDF merge benchmark (#870)
rapidsai/ucx-py: ucp/benchmarks/README.md

# Running Benchmarks

## cuDF Merge

The cuDF merge benchmark can be executed in 3 different ways:

1. Local-node only;
1. Multi-node with automatic process launch via SSH;
1. Multi-node with manual process launch.

In the following subsections we will exemplify how to launch the benchmark for each case. Be sure to check `python -m ucp.benchmarks.cudf_merge --help` to see a complete description of available options and their description.

### Local-node only

> This is the simplest setup and will execute the benchmark only on GPUs where the process is being launched from. In its simple form, it can be executed as:

Reviewer: ```suggestion
This is the simplest setup and will execute the benchmark only using GPUs on the host the process is being launched from. In its simple form, it can be executed as:
```

Once the SSH setup is complete, we can extend the local-node example by simply introducing a new argument `--hosts`. The `--hosts` argument takes a comma-separated list of node hostnames or IP addresses, where the first element is where the server will run, followed by any number of worker hosts. The server is required to synchronize workers and doesn't require a GPU but is allowed to share the same node as one of the worker nodes in that list. For example, if we want the server to run on a node with hostname 'dgx12', with workers on machine with hostname 'dgx12', another with IP address '10.10.10.10' and another with hostname 'dgx13', we would run the command as below:

```bash
python -m ucp.benchmarks.cudf_merge -d 0,1,2,3 -c 1_000_000 --iter 10 --hosts dgx12,dgx12,10.10.10.10,dgx15
```

> Note that in the example above we repeated dgx12, the first node identifies the server, and the remaining arguments identify workers. Multi-node setups also require every node to use the same GPU indices, so if you specify `-d 0,1,2,3`, all worker nodes must have those 4 GPUs available.

Reviewer: ```suggestion
Note that in the example above we repeated dgx12, the first node identifies the server, and the remaining arguments identify workers. Multi-node setups also require every node to use the same GPU indices, so if you specify `-d 0,1,2,3`, all worker nodes must have those four GPUs available.
```

Rationale: spell out numbers below ten.

> Should anything go wrong, you may specify the `UCXPY_ASYNCSSH_LOG_LEVEL=DEBUG` environment variable. This will print additional information, including the exact command that is being executed in server and workers, as well as the output for each process.

Reviewer: ```suggestion
Should anything go wrong, you may specify the `UCXPY_ASYNCSSH_LOG_LEVEL=DEBUG` environment variable. This will print additional information, including the exact command that is being executed on the server and workers, as well as the output for each process.
```

### Multi-node with manual process setup

> This is by far the most complicated setup, and should be used only if running on a cluster where SSH isn't possible or not recommened.

Reviewer: ```suggestion
This is by far the most complicated setup, and should be used only if running on a cluster where SSH isn't possible or not recommended.
```

> Before diving in the details of this setup, the user should know there is a convenience function to generate commands for each of the nodes. To generate that, the user may refer to the SSH section, where we pass `--hosts` to specify server and worker nodes, this time with an added `--print-commands-only` argument. The output of the command will print one line for server and each node with the exact command that should be executed in those nodes.

Reviewer: ```suggestion
Before diving in to the details of this setup, the user should know there is a convenience function to generate commands for each of the nodes. To generate that, the user may refer to the SSH section, where we pass `--hosts` to specify server and worker nodes, this time with an added `--print-commands-only` argument. The output of the command will print one line for server and each node with the exact command that should be executed on those nodes.
```

# chunks_per_dev: number of chunks (processes) per GPU specified via `--chunks-per-dev`
# num_worker_nodes: number of nodes that will run workers
num_workers = len(devs) * chunks_per_dev * num_worker_nodes
```

In the examples that follow, we will launch 4 GPUs per node, 2 chunks per device and 2 worker nodes.

```python
# num_workers = len(devs) * chunks_per_dev * num_worker_nodes
16 = 4 * 2 * 2
```

#### Server

First, the user must launch the server, which may be in a node that doesn't contain any GPUs, or could be on one of the nodes where workers will run. To do so, two options exist:

##### Address on a file

In this setup, a file that is reachable on all nodes of the cluster must be specified, for example in a network file system. 

```bash
python -m ucp.benchmarks.cudf_merge --server --devs 0,1,2,3 --chunks-per-dev 2 --chunk-size 1000000 --frac-match 0.5 --iter 10 --warmup-iter 5 --num-workers 16 --server-file /path/to/network/fs/server.json
```

##### Address on stdout

```bash
python -m ucp.benchmarks.cudf_merge --server --devs 0,1,2,3 --chunks-per-dev 2 --chunk-size 1000000 --frac-match 0.5 --iter 10 --warmup-iter 5 --num-workers 16
```

#### Workers

> Once the server is up and running, workers must be launched on multiple nodes. Each node must execute workers with all options matching other nodes, both workers and server, with the execption of `--node-idx`. The `--node-idx` argument is used as a unique identifier to each worker, so that it may compute the rank for each GPU worker.

Reviewer: I would have expected that `--rmm-init-pool-size` is not a collective specification? I guess that's minor though.

Author: It's not, there are some others too, like `--collect-garbage`, but specifying different options may lead to situations that may be difficult to ensure they indeed work correctly. So for now I think it's best to try and make things homogenous.

#### Server address on a file

```bash

> # Worker 0

Reviewer: Should we indicate that these commands should be executed on the host corresponding to worker 0 and worker 1 respectively?

Author: That was my intent with the `# Worker 0`/`# Worker 1` comments. Do you have a specific preference in mind of how to make this clearer? Just expand on the command or something else?

Reviewer: Maybe `# Worker 0 (run on host_for_worker_0)` `# Worker 1 (run on host_for_worker_1)` ?

Author: I pushed https://github.com/rapidsai/ucx-py/pull/870/commits/21f98fd063652ee483f8295a2687a15c6d882cf3, hopefully that will be in accordance to your request and be a little more descriptive as well.

> /datasets/pentschev/miniconda3/envs/rn-220803/bin/python -m ucp.benchmarks.cudf_merge --devs 0,1,2,3 --chunks-per-dev 2 --chunk-size 1000000 --frac-match 0.5 --iter 10 --warmup-iter 5 --num-workers 16 --node-idx 0 --rmm-init-pool-size 4000000000 --server-file '/path/to/network/fs/server.json'

Reviewer: ```suggestion
python -m ucp.benchmarks.cudf_merge --devs 0,1,2,3 --chunks-per-dev 2 --chunk-size 1000000 --frac-match 0.5 --iter 10 --warmup-iter 5 --num-workers 16 --node-idx 0 --rmm-init-pool-size 4000000000 --server-file '/path/to/network/fs/server.json'
```

For consistency with previous calls

# Worker 1

> /datasets/pentschev/miniconda3/envs/rn-220803/bin/python -m ucp.benchmarks.cudf_merge --devs 0,1,2,3 --chunks-per-dev 2 --chunk-size 1000000 --frac-match 0.5 --iter 10 --warmup-iter 5 --num-workers 16 --node-idx 1 --rmm-init-pool-size 4000000000 --server-file '/path/to/network/fs/server.json'

Reviewer: ```suggestion
python -m ucp.benchmarks.cudf_merge --devs 0,1,2,3 --chunks-per-dev 2 --chunk-size 1000000 --frac-match 0.5 --iter 10 --warmup-iter 5 --num-workers 16 --node-idx 1 --rmm-init-pool-size 4000000000 --server-file '/path/to/network/fs/server.json'
```

```

#### Server address on stdout

```bash
# Worker 0

> /datasets/pentschev/miniconda3/envs/rn-220803/bin/python -m ucp.benchmarks.cudf_merge --devs 0,1,2,3 --chunks-per-dev 2 --chunk-size 1000000 --frac-match 0.5 --iter 10 --warmup-iter 5 --num-workers 16 --node-idx 0 --rmm-init-pool-size 4000000000 --server-address 'REPLACE WITH SERVER ADDRESS'

Reviewer: ```suggestion
python -m ucp.benchmarks.cudf_merge --devs 0,1,2,3 --chunks-per-dev 2 --chunk-size 1000000 --frac-match 0.5 --iter 10 --warmup-iter 5 --num-workers 16 --node-idx 0 --rmm-init-pool-size 4000000000 --server-address 'REPLACE WITH SERVER ADDRESS'
```

# Worker 1

> /datasets/pentschev/miniconda3/envs/rn-220803/bin/python -m ucp.benchmarks.cudf_merge --devs 0,1,2,3 --chunks-per-dev 2 --chunk-size 1000000 --frac-match 0.5 --iter 10 --warmup-iter 5 --num-workers 16 --node-idx 1 --rmm-init-pool-size 4000000000 --server-address 'REPLACE WITH SERVER ADDRESS'

Reviewer: ```suggestion
python -m ucp.benchmarks.cudf_merge --devs 0,1,2,3 --chunks-per-dev 2 --chunk-size 1000000 --frac-match 0.5 --iter 10 --warmup-iter 5 --num-workers 16 --node-idx 1 --rmm-init-pool-size 4000000000 --server-address 'REPLACE WITH SERVER ADDRESS'
```
